File: app/src/main/java/com/tronprotocol/app/guidance/AnthropicApiClient.java
package com.tronprotocol.app.guidance;

import android.text.TextUtils;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.ArrayDeque;
import java.util.Deque;

/**
 * Anthropic API client with auth, rate limiting and robust error handling.
 */
public class AnthropicApiClient {
    public static final String MODEL_SONNET = "claude-sonnet-4-5-20250929";
    public static final String MODEL_OPUS = "claude-opus-4-6";

    private static final String API_URL = "https://api.anthropic.com/v1/messages";
    private static final String API_VERSION = "2023-06-01";

    private final Object rateLock = new Object();
    private final Deque<Long> requestTimes = new ArrayDeque<>();
    private final int maxRequestsPerMinute;
    private final long minRequestIntervalMs;
    private long lastRequestMs = 0;

    public AnthropicApiClient(int maxRequestsPerMinute, long minRequestIntervalMs) {
        this.maxRequestsPerMinute = Math.max(1, maxRequestsPerMinute);
        this.minRequestIntervalMs = Math.max(0, minRequestIntervalMs);
    }

    public String createGuidance(String apiKey, String model, String prompt, int maxTokens) throws AnthropicException {
        if (TextUtils.isEmpty(apiKey)) {
            throw new AnthropicException("Anthropic API key is missing", 401, false);
        }
        if (TextUtils.isEmpty(prompt)) {
            throw new AnthropicException("Prompt is empty", 400, false);
        }

        throttle();

        HttpURLConnection connection = null;
        try {
            URL url = new URL(API_URL);
            connection = (HttpURLConnection) url.openConnection();
            connection.setConnectTimeout(10000);
            connection.setReadTimeout(20000);
            connection.setRequestMethod("POST");
            connection.setDoOutput(true);
            connection.setRequestProperty("Content-Type", "application/json");
            connection.setRequestProperty("x-api-key", apiKey);
            connection.setRequestProperty("anthropic-version", API_VERSION);

            JSONObject body = new JSONObject();
            body.put("model", model);
            body.put("max_tokens", maxTokens);

            JSONArray messages = new JSONArray();
            JSONObject userMessage = new JSONObject();
            userMessage.put("role", "user");
            userMessage.put("content", prompt);
            messages.put(userMessage);
            body.put("messages", messages);

            byte[] payload = body.toString().getBytes(StandardCharsets.UTF_8);
            try (OutputStream os = connection.getOutputStream()) {
                os.write(payload);
            }

            int status = connection.getResponseCode();
            String responseBody = readBody(status >= 400 ? connection.getErrorStream() : connection.getInputStream());

            if (status >= 400) {
                throw buildError(responseBody, status);
            }

            JSONObject responseJson = new JSONObject(responseBody);
            JSONArray content = responseJson.optJSONArray("content");
            if (content == null || content.length() == 0) {
                throw new AnthropicException("Anthropic response did not contain content", status, true);
            }

            StringBuilder answer = new StringBuilder();
            for (int i = 0; i < content.length(); i++) {
                JSONObject block = content.optJSONObject(i);
                if (block != null && "text".equals(block.optString("type"))) {
                    answer.append(block.optString("text", ""));
                }
            }

            if (answer.length() == 0) {
                throw new AnthropicException("Anthropic response text was empty", status, true);
            }

            return answer.toString().trim();
        } catch (AnthropicException e) {
            throw e;
        } catch (Exception e) {
            throw new AnthropicException("Failed to call Anthropic API: " + e.getMessage(), 500, true);
        } finally {
            if (connection != null) {
                connection.disconnect();
            }
        }
    }

    private void throttle() throws AnthropicException {
        synchronized (rateLock) {
            long now = System.currentTimeMillis();

            while (!requestTimes.isEmpty() && now - requestTimes.peekFirst() > 60000) {
                requestTimes.pollFirst();
            }

            if (requestTimes.size() >= maxRequestsPerMinute) {
                throw new AnthropicException("Rate limit hit in client guardrail", 429, true);
            }

            long sinceLast = now - lastRequestMs;
            if (sinceLast < minRequestIntervalMs) {
                long wait = minRequestIntervalMs - sinceLast;
                try {
                    Thread.sleep(wait);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    throw new AnthropicException("Rate limiter interrupted", 429, true);
                }
            }

            long stamp = System.currentTimeMillis();
            requestTimes.addLast(stamp);
            lastRequestMs = stamp;
        }
    }

    private AnthropicException buildError(String body, int status) {
        String msg = "Anthropic API error";
        try {
            JSONObject json = new JSONObject(body);
            JSONObject error = json.optJSONObject("error");
            if (error != null) {
                msg = error.optString("message", msg);
            }
        } catch (Exception ignored) {
            if (!TextUtils.isEmpty(body)) {
                msg = body;
            }
        }

        boolean retryable = status == 429 || status >= 500;
        return new AnthropicException(msg, status, retryable);
    }

    private String readBody(InputStream inputStream) throws Exception {
        if (inputStream == null) {
            return "";
        }
        try (BufferedReader reader = new BufferedReader(
            new InputStreamReader(inputStream, StandardCharsets.UTF_8))) {
            StringBuilder sb = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                sb.append(line);
            }
            return sb.toString();
        }
    }

    public static class AnthropicException extends Exception {
        private final int statusCode;
        private final boolean retryable;

        public AnthropicException(String message, int statusCode, boolean retryable) {
            super(message);
            this.statusCode = statusCode;
            this.retryable = retryable;
        }

        public int getStatusCode() {
            return statusCode;
        }

        public boolean isRetryable() {
            return retryable;
        }
    }
}

File: app/src/main/java/com/tronprotocol/app/guidance/AnthropicApiClient.kt
package com.tronprotocol.app.guidance

import android.text.TextUtils
import org.json.JSONArray
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStream
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import java.nio.charset.StandardCharsets
import java.util.ArrayDeque

/**
 * Anthropic API client with auth, rate limiting and robust error handling.
 */
class AnthropicApiClient(
    maxRequestsPerMinute: Int,
    minRequestIntervalMs: Long
) {

    private val rateLock = Any()
    private val requestTimes = ArrayDeque<Long>()
    private val maxRequestsPerMinute: Int = maxOf(1, maxRequestsPerMinute)
    private val minRequestIntervalMs: Long = maxOf(0L, minRequestIntervalMs)
    private var lastRequestMs: Long = 0L

    @Throws(AnthropicException::class)
    fun createGuidance(apiKey: String?, model: String, prompt: String?, maxTokens: Int): String {
        if (TextUtils.isEmpty(apiKey)) {
            throw AnthropicException("Anthropic API key is missing", 401, false)
        }
        if (TextUtils.isEmpty(prompt)) {
            throw AnthropicException("Prompt is empty", 400, false)
        }

        throttle()

        var connection: HttpURLConnection? = null
        try {
            val url = URL(API_URL)
            connection = url.openConnection() as HttpURLConnection
            connection.connectTimeout = 10000
            connection.readTimeout = 20000
            connection.requestMethod = "POST"
            connection.doOutput = true
            connection.setRequestProperty("Content-Type", "application/json")
            connection.setRequestProperty("x-api-key", apiKey)
            connection.setRequestProperty("anthropic-version", API_VERSION)

            val body = JSONObject().apply {
                put("model", model)
                put("max_tokens", maxTokens)

                val messages = JSONArray()
                val userMessage = JSONObject().apply {
                    put("role", "user")
                    put("content", prompt)
                }
                messages.put(userMessage)
                put("messages", messages)
            }

            val payload = body.toString().toByteArray(StandardCharsets.UTF_8)
            connection.outputStream.use { os ->
                os.write(payload)
            }

            val status = connection.responseCode
            val responseBody = readBody(
                if (status >= 400) connection.errorStream else connection.inputStream
            )

            if (status >= 400) {
                throw buildError(responseBody, status)
            }

            val responseJson = JSONObject(responseBody)
            val content = responseJson.optJSONArray("content")
            if (content == null || content.length() == 0) {
                throw AnthropicException("Anthropic response did not contain content", status, true)
            }

            val answer = StringBuilder()
            for (i in 0 until content.length()) {
                val block = content.optJSONObject(i)
                if (block != null && "text" == block.optString("type")) {
                    answer.append(block.optString("text", ""))
                }
            }

            if (answer.isEmpty()) {
                throw AnthropicException("Anthropic response text was empty", status, true)
            }

            return answer.toString().trim()
        } catch (e: AnthropicException) {
            throw e
        } catch (e: Exception) {
            throw AnthropicException("Failed to call Anthropic API: ${e.message}", 500, true)
        } finally {
            connection?.disconnect()
        }
    }

    @Throws(AnthropicException::class)
    private fun throttle() {
        synchronized(rateLock) {
            var now = System.currentTimeMillis()

            while (requestTimes.isNotEmpty()) {
                val oldest = requestTimes.peekFirst() ?: break
                if (now - oldest <= 60000) break
                requestTimes.pollFirst()
                now = System.currentTimeMillis()
            }

            if (requestTimes.size >= maxRequestsPerMinute) {
                throw AnthropicException("Rate limit hit in client guardrail", 429, true)
            }

            val sinceLast = now - lastRequestMs
            if (sinceLast < minRequestIntervalMs) {
                val wait = minRequestIntervalMs - sinceLast
                try {
                    Thread.sleep(wait)
                } catch (e: InterruptedException) {
                    Thread.currentThread().interrupt()
                    throw AnthropicException("Rate limiter interrupted", 429, true)
                }
            }

            val stamp = System.currentTimeMillis()
            requestTimes.addLast(stamp)
            lastRequestMs = stamp
        }
    }

    private fun buildError(body: String, status: Int): AnthropicException {
        var msg = "Anthropic API error"
        try {
            val json = JSONObject(body)
            val error = json.optJSONObject("error")
            if (error != null) {
                msg = error.optString("message", msg)
            }
        } catch (ignored: Exception) {
            if (!TextUtils.isEmpty(body)) {
                msg = body
            }
        }

        val retryable = status == 429 || status >= 500
        return AnthropicException(msg, status, retryable)
    }

    @Throws(Exception::class)
    private fun readBody(inputStream: InputStream?): String {
        if (inputStream == null) {
            return ""
        }
        BufferedReader(InputStreamReader(inputStream, StandardCharsets.UTF_8)).use { reader ->
            val sb = StringBuilder()
            var line: String?
            while (reader.readLine().also { line = it } != null) {
                sb.append(line)
            }
            return sb.toString()
        }
    }

    class AnthropicException(
        message: String,
        val statusCode: Int,
        val isRetryable: Boolean
    ) : Exception(message)

    companion object {
        const val MODEL_SONNET = "claude-sonnet-4-5-20250929"
        const val MODEL_OPUS = "claude-opus-4-6"

        private const val API_URL = "https://api.anthropic.com/v1/messages"
        private const val API_VERSION = "2023-06-01"
    }
}
